File: template_to_pdf/extract.py
import json
import re

from .schema import normalize

PROMPT = 'Extract JSON {"title": string, "sections": [{"heading": string, "body": string}], "meta"?: object} from the text. Reply with JSON only.'

_FENCE = re.compile(r"```(?:json)?\s*([\s\S]*?)```")


def _parse_json(text):
    if not isinstance(text, str):
        return None
    text = text.strip()
    try:
        val = json.loads(text)
        if isinstance(val, dict):
            return val
    except json.JSONDecodeError:
        pass
    for m in _FENCE.finditer(text):
        try:
            val = json.loads(m.group(1).strip())
            if isinstance(val, dict):
                return val
        except json.JSONDecodeError:
            continue
    return None


def extract(text, llm=None):
    data = _parse_json(text)
    if data is None:
        if llm is None:
            raise ValueError("non-JSON input needs llm")
        data = _parse_json(llm(PROMPT + "\n\n" + text))
        if data is None:
            raise ValueError("llm did not return JSON")
    return normalize(data)

File: js/extract.js
import { normalize } from "./schema.js"

export const PROMPT =
  'Extract JSON {"title": string, "sections": [{"heading": string, "body": string}], "meta"?: object} from the text. Reply with JSON only.'

export function parseJson(text) {
  if (typeof text !== "string") return null
  text = text.trim()
  try {
    const val = JSON.parse(text)
    if (val && typeof val === "object" && !Array.isArray(val)) return val
  } catch {
    // try code fences
  }
  const fenceRegex = /```(?:json)?\s*([\s\S]*?)```/g
  let m
  while ((m = fenceRegex.exec(text)) !== null) {
    try {
      const val = JSON.parse(m[1].trim())
      if (val && typeof val === "object" && !Array.isArray(val)) return val
    } catch {
      // try next block
    }
  }
  return null
}

export function extract(text, llm) {
  let data = parseJson(text)
  if (data == null) {
    if (!llm) throw new Error("non-JSON input needs llm")
    const res = llm(PROMPT + "\n\n" + text)
    if (res && typeof res.then === "function") {
      return res.then((resolved) => {
        const d = parseJson(resolved)
        if (d == null) throw new Error("llm did not return JSON")
        return normalize(d)
      })
    }
    data = parseJson(res)
    if (data == null) throw new Error("llm did not return JSON")
  }
  return normalize(data)
}
